io/src/lib.rs

```diff
@@ -143,35 +143,32 @@ impl<'a, R: BufRead + ?Sized> BufRead for Take<'a, R> {
     }
 }
 
-#[cfg(feature = "std")]
-impl<R: std::io::Read> Read for R {
+impl Read for &[u8] {
     #[inline]
     fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
-        Ok(<R as std::io::Read>::read(self, buf)?)
+        let cnt = cmp::min(self.len(), buf.len());
+        buf[..cnt].copy_from_slice(&self[..cnt]);
+        *self = &self[cnt..];
+        Ok(cnt)
     }
 }
 
 #[cfg(feature = "std")]
-impl<R: std::io::BufRead + Read + ?Sized> BufRead for R {
+impl Read for std::net::TcpStream {
     #[inline]
-    fn fill_buf(&mut self) -> Result<&[u8]> { Ok(std::io::BufRead::fill_buf(self)?) }
-
-    #[inline]
-    fn consume(&mut self, amount: usize) { std::io::BufRead::consume(self, amount) }
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        Ok(std::io::Read::read(self, buf)?)
+    }
 }
 
-#[cfg(not(feature = "std"))]
-impl Read for &[u8] {
+#[cfg(feature = "std")]
+impl Read for std::io::BufReader<std::net::TcpStream> {
```

Comment:
This needs to be `T: std::io::Read`, no point implementing it for `TcpStream` only.

Reply:
I had that at first but we are specifically trying to remove the blanket impls so I did this?

Reply:
`impl<T: std::io::Read> for std::io::BufReader<T>` is not really a blanket impl. Or at least not in the sense that would break things.

Reply:
Right -- `BufReader` is a "concrete type" even though it's parameterized.